gobblin-api/src/main/java/gobblin/source/workunit/ExtractFactory.java

```diff
@@ -0,0 +1,49 @@
+package gobblin.source.workunit;
+
+import java.util.Locale;
+import java.util.Set;
+
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.Sets;
+
+import gobblin.source.workunit.Extract.TableType;
+
+
+public class ExtractFactory {
+  private final Set<Extract> createdInstances;
+  private final DateTimeFormatter dtf;
+
+  public ExtractFactory(String dateTimeFormat) {
+    this.createdInstances = Sets.newConcurrentHashSet();
+    this.dtf = DateTimeFormat.forPattern(dateTimeFormat).withLocale(Locale.US).withZone(DateTimeZone.UTC);
+  }
+
+  /**
+   * Returns a unique {@link Extract} instance.
+   * Any two calls of this method from the same {@link ExtractFactory} instance guarantees to
+   * return {@link Extract}s with different IDs.
+   *
+   * @param type {@link TableType}
+   * @param namespace dot separated namespace path
+   * @param table table name
+   * @return a unique {@link Extract} instance
+   */
+  public Extract getUniqueExtract(TableType type, String namespace, String table) {
```

Comment:
This needs to be synchronized and the createdInstances need not. There are race conditions in the current implementation.